core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemWriter.java

```diff
@@ -0,0 +1,35 @@
+package org.apache.spark.util.collection.unsafe.sort;
+
+import org.apache.spark.executor.ShuffleWriteMetrics;
+import org.apache.spark.memory.TaskMemoryManager;
+import org.apache.spark.unsafe.Platform;
+import org.apache.spark.unsafe.memory.MemoryBlock;
+
+import java.util.LinkedList;
+
+public final class PMemWriter {
+    private final ShuffleWriteMetrics writeMetrics;
+    private final TaskMemoryManager taskMemoryManager;
+    private int numRecordsWritten = 0;
+    private final LinkedList<MemoryBlock> allocatedPMemPages = new LinkedList<>();
+
+    public PMemWriter(
+            ShuffleWriteMetrics writeMetrics,
+            TaskMemoryManager taskMemoryManager) {
+        this.writeMetrics = writeMetrics;
+        this.taskMemoryManager = taskMemoryManager;
+    }
+    public boolean dumpPageToPMem(MemoryBlock page, int numRecordsInPage) {
+        MemoryBlock pMemBlock = taskMemoryManager.allocatePMemPage(page.size() + 4);
+        if (pMemBlock != null) {
+            Platform.putInt(null, pMemBlock.getBaseOffset(), numRecordsInPage);
+            Platform.copyMemory(page.getBaseObject(), page.getBaseOffset(), null, pMemBlock.getBaseOffset(), page.size());
```

Comment:
Here you put `numRecordsInPage`  to the first 4 bytes, so when copy memory, need to add this 4 bytes, right?

i mean it should be `pMemBlock.getBaseOffset() + 4` @yma11

Reply:
Yes. This is a bug. I found it in my UT. :)